src/MongoDB.Driver.Core/Core/Misc/StreamExtensionMethods.cs

```diff
@@ -36,6 +36,39 @@ public static void EfficientCopyTo(this Stream input, Stream output)
             }
         }
 
+        public static async Task<int> ReadAsync(this Stream stream, byte[] buffer, int offset, int count, TimeSpan timeout, CancellationToken cancellationToken)
+        {
+            var state = 1; // 1 == reading, 2 == done reading, 3 == timedout, 4 == cancelled
+            void changeState(int to)
+            {
+                var from = Interlocked.CompareExchange(ref state, to, 1);
+                if (from == 1 && to >= 3)
+                {
+                    try { stream.Dispose(); } catch { } // disposing the stream aborts the read attempt
+                }
+            }
+
+            var bytesRead = 0;
+            using (new Timer(_ => changeState(3), null, timeout, Timeout.InfiniteTimeSpan))
+            using (cancellationToken.Register(() => changeState(4)))
+            {
+                try
+                {
+                    bytesRead = await stream.ReadAsync(buffer, offset, count, cancellationToken).ConfigureAwait(false);
+                    changeState(2);
```

Comment:
State 2 is necessary to prevent the race condition where we RETURN TO THE USER with the appearance of having Read successfully BUT while also having disposed  of the stream because of a timeout or cancellation request.